### DISCUSSION TITLE
Can you use keystore entries in task templates

### DISCUSSION THREAD

**Discussion by smitch35** [semaphoreui/semaphore#1361] [Q&A] [answered]
So my co worker and I are fairly new to ansible and ansible semaphore but we've have some things working and they work well.  One thing we're trying to do is pass usernames and passwords to certain things when accessing them via ansible (for example vmware vcenter).

We know we could leave some creds in the playbook or reference a vault but we'd like to be able to store a username and password in the keystore and then use them in variables for tasks.  Is this possible at all or not really intended purpose?

Thanks,

**Comment by ansibleguy** [comment, marked as answer]
Greetings @smitch35 
As far as I have seen - the keystore can only be used to supply the credentials to run tasks and clone repositories.
Essentially what you want seems to be to pass `extra-vars` with secret values to Ansible so you can switch secrets on a per-execution basis, right?

Practical usage of some use-cases like these used a 'secrets.yml' either in the global 'vars' directory (_placement in the same depth as your playbooks_) or in a role-specific vars-file.
See the documentation for [directory structure](https://docs.ansible.com/ansible/2.8/user_guide/playbooks_best_practices.html#directory-layout) and [search paths](https://docs.ansible.com/ansible/latest/playbook_guide/playbook_pathing.html#resolving-local-relative-paths).

Example 'secrets.yml':
```yaml
vmware_a:
  selector1: !vault |
    ...
  selector2: !vault |
    ...
  selector3: !vault |
    ...
# NOTE: 'selectorN' is just a generic placeholder
```

Example 'task':
```yaml
- debug:
     var: vmware_a[run_selector]
```

Execution:
```bash
ansible-playbook play.yml --extra-vars run_selector=selector2 --ask-vault-pass
```

\- AnsibleGuy

**Reply by smitch35** [reply]
Thanks for this information.  I was just looking at it a bit wrong.  So if I'm understanding this correctly.  I need to locally access my system running semaphore to build out my vault and then reference those vault entries because they need to be outside of semaphore but I can put the vault password in semaphore to pass along to the script when running?  

I assume I could also make a vault and include it in my repo but I'm still learning some aspects of maintaining a repo etc.  

Thanks for the answer!

**Reply by ansibleguy** [reply]
Exactly.
You can add 'ansible-vault' variables in your repository scope or include var-files from anywhere else in the filesystem using 'vars_files' (inside playbook) or '[include_vars](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/include_vars_module.html)' (inside tasks/roles)
